src/posit/connect/paginator.py

```diff
@@ -1,49 +1,111 @@
+        self.exhausted = False
+        self.index = 0
+        self.page_size = page_size
         self.session = session
         self.url = url
-        self.page_number = start_page
-        self.page_size = page_size
-        # The API response will tell us how many total entries there are,
-        # but we don't know yet.
-        self.total = None
-        # We also want to track how many results we've seen so far
-        self.seen = 0
-
-    def get_all(self) -> List[dict]:
-        result = []
-        while self.total is None or self.seen < self.total:
-            result += self.get_next_page()
-            self.page_number += 1
-        return result
-
-    def get_next_page(self) -> List[dict]:
+
+    def fetch(self, page: int) -> Tuple[Iterator[dict], bool]:
+        """
+        Fetches a specific page of results from the API.
+
+        Args:
+            page (int): The page number to fetch. Must be greater than 0.
+
+        Returns:
+            tuple[Iterator[dict] | None, bool]: A tuple containing the fetched results and a boolean indicating
+            whether there are more pages to fetch.
+
+        Raises:
+            ValueError: If the page number is not greater than 0.
+        """
+        # Check if page is greater than 0
+        if not (page > 0):
+            raise ValueError(f"page must be greater than 0, got {page}")
         # Define query parameters for pagination.
-        params = {"page_number": self.page_number, "page_size": self.page_size}
+        params = {"page_number": page, "page_size": self.page_size}
         # Send a GET request to the endpoint with the specified parameters.
-        response = self.session.get(self.url, params=params).json()
-        # On our first request, we won't have set the total yet, so do it
-        if self.total is None:
-            self.total = response["total"]
-        results = response["results"]
-        self.seen += len(results)
-        return results
+        response = self.session.get(self.url, params=params)
+        # Convert response to dict
+        as_dict: dict = dict(response.json())
+        # Parse the JSON response and extract the results.
+        results = as_dict["results"]
+        # Mark exhausted if the result size is less than the maximum page size.
+        exhausted = len(results) < self.page_size
+        return (results, exhausted)
+
+    def __iter__(self) -> Iterator[dict]:
+        """
+        Returns an iterator object for the Paginator class.
+
+        Returns:
+            Iterator[dict]: An iterator object that iterates over the items in the Paginator.
+        """
+        self.index = 0
+        return self
+
+    def __len__(self) -> int:
+        """
+        Returns the total number of items in the paginator.
+
+        Returns:
+            int: The total number of items in the paginator.
+        """
+        # Define query parameters for pagination.
+        params = {"page_number": 1, "page_size": 1}
```

Comment:
Does `page_size=0` work?

Reply:
Unfortunately, no, `posit.connect.errors.ClientError: The parameter page_size=0 is less than the allowed minimum 1 (Error Code: 25, HTTP Status: 400 Bad Request)`.